amaranth/vendor/xilinx.py

```diff
@@ -411,6 +429,89 @@ def _part(self):
         """
     ]
 
+    # Yosys NextPNR prjxray templates
+
+    @property
+    def _xray_part_map(self):
+        return {
+            "xc7a35ticsg324-1L":  "xc7a35tcsg324-1",  # Arty-A7 35t
+            "xc7a100ticsg324-1L": "xc7a100tcsg324-1", # Arty-A7 100t
+        }.get(self._part, self._part)
+
+    _xray_device = {
+        "xc7a35ti":  "xc7a35t",
+        "xc7a100ti": "xc7a100t",
+    }
+
+    _xray_family = {
+        "xc7a": "artix7",
+        "xc7z": "zynq7",
+    }
+
+    _xray_required_tools = [
+        "yosys",
+        "nextpnr-xilinx",
+        "fasm2frames",
+        "xc7frames2bit"
+    ]
+    _xray_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "build_{{name}}.sh": r"""
+            # {{autogenerated}}
+            set -e{{verbose("x")}}
+            [ -n "${{platform._deprecated_toolchain_env_var}}" ] && . "${{platform._deprecated_toolchain_env_var}}"
+            [ -n "${{platform._toolchain_env_var}}" ] && . "${{platform._toolchain_env_var}}"
+            : ${DB_DIR:={{get_override("nextpnr_dir")|default("/usr/share/nextpnr")}}/prjxray-db}
+            : ${CHIPDB_DIR:={{get_override("nextpnr_dir")|default("/usr/share/nextpnr")}}/xilinx-chipdb}
+            {{emit_commands("sh")}}
+        """,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.xdc": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                {% for attr_name, attr_value in attrs.items() -%}
+                    set_property {{attr_name}} {{attr_value}} [get_ports {{port_name|tcl_escape}}]
+                    set_property LOC {{pin_name}} [get_ports {{port_name|tcl_escape}}]
+                {% endfor %}
+            {% endfor %}
+        """,
+    }
+    _xray_command_templates = [
+        r"""
+        {{invoke_tool("yosys")}}
+            -p "synth_xilinx -flatten -abc9 -nobram -arch xc7 -top {{name}}; write_json {{name}}.json"
+            {% for file in platform.iter_files(".v", ".sv", ".vhd", ".vhdl") -%} {{file}} {% endfor %}
+            {{name}}.v
+        """,
+        r"""
+        {{invoke_tool("nextpnr-xilinx")}}
+            --chipdb $CHIPDB_DIR/{{platform._xray_device.get(platform.device, platform.device)}}.bin
+            --xdc {{name}}.xdc
+            --json {{name}}.json
+            --write {{name}}_routed.json
+            --fasm {{name}}.fasm
+        """,
+        r"""
+        {{invoke_tool("fasm2frames")}}
+            --part {{platform._xray_part_map}}
+            --db-root $DB_DIR/{{platform._xray_family[platform.device[:4]]}} {{name}}.fasm > {{name}}.frames
```

Comment:
The whole `platform._xray_family[platform.device[:4]]` subexpression, and the dict `platform._xray_family` as well, could go into an `_xray_family` property.